WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift

```diff
@@ -19,11 +19,683 @@
+            let originalText = clientMessage.underlyingMessage?.textData,
+            case .text? = messageEdit.content,
+            senderID == clientMessage.sender?.remoteIdentifier
+        else {
+            return false
+        }
+
+        do {
+            let genericMessage = GenericMessage(
+                content: originalText.applyEdit(from: messageEdit.text),
+                nonce: messageNonce
+            )
+            try clientMessage.setUnderlyingMessage(genericMessage)
+        } catch {
+            WireLogger.messageProcessing.warn(
+                "Failed to process message edit. Reason: \(error.localizedDescription)"
+            )
+            return false
+        }
+
+        clientMessage.updateNormalizedText()
+        clientMessage.nonce = messageNonce
+        clientMessage.updatedTimestamp = date
+        clientMessage.reactions.removeAll()
+        clientMessage.linkAttachments = nil
+
+        return true
+    }
+    
+    private func insertTextMessage(
+        _ message: GenericMessage,
+        conversation: ZMConversation,
+        senderID: UUID,
+        senderDomain: String,
+        senderClientID: String?,
+        date: Date?,
+        logAttributes: LogAttributes
+    ) {
+        
+        func shouldAdd() -> Bool {
+            if let clearedTime = conversation.clearedTimeStamp, let time = date,
+               clearedTime.compare(time) != .orderedAscending {
+                return false
+            }
+            return conversation.conversationType != .self
+        }
+        
+        guard shouldAdd(), let nonce = UUID(uuidString: message.messageID) else {
+            return WireLogger.eventProcessing.warn(
+                "Dropping message because no nonce or for self conv",
+                attributes: logAttributes
+            )
+        }
+        
+        let messageClass: AnyClass = GenericMessage.entityClass(for: message)
+        
+        var clientMessage = messageClass.fetch(
+            withNonce: nonce,
+            for: conversation,
+            in: context,
+            prefetchResult: .none
+        ) as? ZMOTRMessage
+        
+        func isZombieObject(_ message: ZMOTRMessage?) -> Bool {
+            guard let message else { return false }
+            return message.isZombieObject
+        }
+        
+        guard !isZombieObject(clientMessage)  else {
+            return WireLogger.eventProcessing.warn(
+                "Dropping message because zombieObject",
+                attributes: logAttributes
+            )
+        }
+        
+        var isNewMessage = false
+        
+        if clientMessage == nil {
+            isNewMessage = true
+            if messageClass is ZMClientMessage.Type {
+                clientMessage = ZMClientMessage(
+                    nonce: nonce,
+                    managedObjectContext: context
+                )
+                
+            } else if messageClass is ZMAssetClientMessage.Type {
```

Comment:
This method is quite big because the same code is used to handle multiple events, typically here I assume this line is meant to handle the `"conversation.otr-asset-add"` . I kept it but since we don't seem to handle this proteus asset add event anymore maybe it could be removed..